docs/source/basic/library.rst

```diff
@@ -115,6 +115,68 @@ The ``alpaka::getWorkDiv`` and the ``alpaka::getIdx`` functions both return a ve
 They are parametrized by the origin of the calculation as well as the unit in which the values are calculated.
 For example, ``alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc)`` returns a vector with the extents of the grid in units of threads.
 
+Memory fences
+`````````````
+
+**Note**: Memory fences should not be mistaken for synchronization functions between threads. They solely enforce the
+ordering of certain memory operations (see below) and how other threads observe this order. If you need to rely on the
+results of memory operations being visible to other threads please use ``alpaka::syncBlockThreads`` or atomic functions
+instead.
```

Comment:
I don't think the wording is fully clear here, because a fence *is* used to influence the order of how changes from one thread become visible to another thread. I would add some emphasis:

```suggestion
**Note**: Memory fences should not be mistaken for synchronization functions between threads.
They solely enforce the ordering of certain memory operations (see below) and restrict the possibilities of how other threads *can* observe this order.
If you need to guarantee that the results of memory operations are visible to other threads you must use ``alpaka::syncBlockThreads`` or atomic functions instead.
```

Reply:
I don't really see a difference between those sentences, sorry. I'm also not certain that fences restrict "the possibilities of how other threads can observe this order" since there aren't any possibilities left once a fence is issued.

Reply:
>I'm also not certain that fences restrict "the possibilities of how other threads can observe this order" since there aren't any possibilities left once a fence is issued.

Of course there are possibilities:

```
int a = 0, b = 0;
void thread1() {
    a = 1;
    fence();
    b = 3;
}
void thread2() {
    read(a);
    fence();
    read(b);
}
```

Thread 2 can observe a=0/b=0 or a=1/b=0 or a=1/b=3. However, the possibility of a=0/b=3 is not possible because of the fence. The fence thus restricts the 4 possible variable assignments to 3. The behavior of thread 2 is not deterministic by the fence. That is also why I wrote that thread 2 **can** observe the behavior, because in the case of a=0/b=0 it does not observe the changes from thread 1.

Reply:
You're talking about other threads observing different **results**, but the **order of instructions** is still the same. With the fence it is guaranteed that a is written before b, but the observed results are undefined.

Reply:
I see. Then please rephrase at least:
```suggestion
**Note**: Memory fences should not be mistaken for synchronization functions between threads. They solely enforce the
ordering of certain memory instructions (see below) and how other threads can observe this order. If you need to rely on the
results of memory operations being visible to other threads please use ``alpaka::syncBlockThreads`` or atomic functions
instead.
```
This way we are specific about instructions. And a fence does not enforce how other threads observe results, it merely restricts what observations are possible.